lib/model.js

```diff
@@ -1050,130 +1051,116 @@ Connection = AmpersandModel.extend({
-        attrs.kerberosPassword = password;
-      } else if (attrs.authStrategy === 'MONGODB') {
-        attrs.mongodbUsername = user;
-        attrs.mongodbPassword = password;
-
-        // authSource takes precedence, but fall back to admin
-        // @note Durran: This is not the documented behaviour of the connection string
-        // but the shell also does not fall back to the dbName and will use admin.
-        try {
-          attrs.mongodbDatabaseName = decodeURIComponent(
-            attrs.authSource || Connection.MONGODB_DATABASE_NAME_DEFAULT,
-            callback
-          );
-        } catch (decodeError) {
-          return callback(decodeError);
-        }
+    if (attrs.authStrategy === 'LDAP') {
+      attrs.ldapUsername = user;
+      attrs.ldapPassword = password;
+    } else if (attrs.authStrategy === 'X509') {
+      attrs.x509Username = user;
+    } else if (attrs.authStrategy === 'KERBEROS') {
+      attrs.kerberosPrincipal = user;
+      attrs.kerberosPassword = password;
+    } else if (attrs.authStrategy === 'MONGODB') {
+      attrs.mongodbUsername = user;
+      attrs.mongodbPassword = password;
+
+      // authSource takes precedence, but fall back to admin
+      // @note Durran: This is not the documented behaviour of the connection string
+      // but the shell also does not fall back to the dbName and will use admin.
+      attrs.mongodbDatabaseName = decodeURIComponent(
+        attrs.authSource || Connection.MONGODB_DATABASE_NAME_DEFAULT
+      );
 
-        Object.assign(
-          attrs,
-          Connection._improveAtlasDefaults(url, attrs.auth.password, attrs.ns)
-        );
-      }
+      Object.assign(
+        attrs,
+        Connection._improveAtlasDefaults(url, attrs.auth.password, attrs.ns)
+      );
     }
+  }
 
-    try {
-      const connection = new Connection(attrs);
-      return callback(null, connection);
-    } catch (error) {
-      return callback(error);
-    }
-  });
+  return new Connection(attrs);
+}
+
+/**
+ * For easy command line integration.
+ *
+ * @example
+ *   const args = require('minimist')(process.argv.slice(2));
+ *   const Connection = require('mongodb-connection-model');
+ *   const createClient = require('scout-client');
+ *   args.endpoint = args.endpoint || 'https://localhost:29017';
+ *   const client = createClient(args.endpoint, Connection.from(url, () => {}));
+ *
+ * @param {String} [url]
+ * @param {Function} [callback]
+ */
+Connection.from = (url, callback) => {
+  try {
+    createConnectionFromUrl(url)
+      .then((connection) => callback(null, connection))
+      .catch(
+        // setImmediate will execute the callback in a different asynchronous context
+        // from the promise rejection. In this way if the callback throws
+        // it would result in an unhandled exception rather that an unhandled
+        // promise rejection which wouldn't be expected with a callback.
+        (error) => setImmediate(() => callback(error))
+      );
```

Comment:
If `createConnectionFromUrl()` succeeds but the callback that we invoke from `.then()` throws, this would end up here and call `callback` a second time … probably not what we want? Maybe we should defer the callback with `setImmediate()` in both cases to avoid that?

Reply:
makes sense, would swapping `.catch` with `.then` do the same?

Reply:
That would then lead to the `.then()` handler always being called, even if the callback was already called in `.then()` … it’s probably easiest to just use the 2-argument version of then and not use `.catch()` this way at all

Reply:
Or we could use `setImmediate()` in the `.then()` handler as well to avoid catching rejections from there

Reply:
Nah then it would complain about unhandled rejection again, never mind, great catch!

Reply:
Oh btw: Nah then it would complain about unhandled rejection again, never mind, great catch!

Was referred to my own comment not yours 😄